src/table_builder.cc

```diff
@@ -87,35 +116,43 @@ void TitanTableBuilder::Add(const Slice& key, const Slice& value) {
       // doing level merge.
       if (get_status.ok()) {
         std::string index_value;
-        AddBlob(ikey.user_key, record.value, &index_value);
+        BlobRecord blob_record;
+        blob_record.key = ikey.user_key;
+        blob_record.value = record.value;
+        BlobFileBuilder::OutContexts contexts;
+        AddBlob(blob_record, ikey, &contexts);
+        AddToBaseTable(contexts);
         UpdateIOBytes(prev_bytes_read, prev_bytes_written, &io_bytes_read_,
                       &io_bytes_written_);
-        if (ok()) {
-          std::string index_key;
-          ikey.type = kTypeBlobIndex;
-          AppendInternalKey(&index_key, ikey);
-          base_builder_->Add(index_key, index_value);
-          return;
-        }
+        if (ok()) return;
       } else {
         ++error_read_cnt_;
         ROCKS_LOG_DEBUG(db_options_.info_log,
                         "Read file %" PRIu64 " error during level merge: %s",
                         index.file_number, get_status.ToString().c_str());
       }
     }
-    base_builder_->Add(key, value);
+    if (builder_unbuffered()) {
+      base_builder_->Add(key, value);
+    } else {
+      std::unique_ptr<BlobFileBuilder::BlobRecordContext> ctx =
+          NewCachedRecordContext(ikey, value);
+      blob_builder_->CacheContext(std::move(ctx));
+    }
   } else {
+    assert(builder_unbuffered());
     base_builder_->Add(key, value);
   }
 }
 
-void TitanTableBuilder::AddBlob(const Slice& key, const Slice& value,
-                                std::string* index_value) {
+void TitanTableBuilder::AddBlob(const BlobRecord& record,
```

Comment:
```
void TitanTableBuilder::AddBlob(
    const Slice& key, const Slice& value,
    const ParsedInternalKey& ikey,
    BlobFileBuilder::OutContexts* contexts) {
```
and construct `BlobRecord` inside the method?

Reply:
I found out actually 

```
void TitanTableBuilder::AddBlob(const ParsedInternalKey& ikey,
                                const Slice& value,
                                BlobFileBuilder::OutContexts* contexts)
```
is enough, key can be read from `ikey.user_key`